src/parser/sparqlParser/SparqlQleverVisitor.cpp

```diff
@@ -198,22 +203,182 @@ std::optional<Values> Visitor::visitTypesafe(Parser::ValuesClauseContext* ctx) {
+  }
+  return {as, bs};
+}
+
+template <typename A, typename B>
+std::pair<vector<A>, vector<B>> splitUp(vector<std::pair<A, B>> elems) {
+  vector<A> as;
+  vector<B> bs;
+  for (auto [a, b] : elems) {
+    as.push_back(a);
+    bs.pusb_back(b);
+  }
+  return {as, bs};
+}
+}  // namespace
+
+std::pair<vector<GraphPatternOperation>, vector<SparqlFilter>>
+Visitor::visitTypesafe(Parser::GroupGraphPatternSubContext* ctx) {
+  // TODO: Make TripleComponent constructible from these types.
+  auto iri = [](const Iri& iri) -> TripleComponent { return iri.toSparql(); };
+  auto blankNode = [](const BlankNode& blankNode) -> TripleComponent {
+    return blankNode.toSparql();
+  };
+  auto literal = [](const Literal& literal) {
+    // Problem: ql:contains-word causes the " to be stripped.
+    // TODO: Move stripAndLowercaseKeywordLiteral out to this point or
+    //  rewrite the Turtle Parser s.t. this code can be integrated into the
+    //  visitor. In this case the turtle parser should output the
+    //  corresponding modell class.
+    try {
+      return TurtleStringParser<TokenizerCtre>::parseTripleObject(
+          literal.toSparql());
+    } catch (const TurtleStringParser<TokenizerCtre>::ParseException&) {
+      return TripleComponent{literal.toSparql()};
+    }
+  };
+  auto graphTerm = [&iri, &blankNode, &literal](const GraphTerm& term) {
+    return term.visit(
+        ad_utility::OverloadCallOperator{iri, blankNode, literal});
+  };
+  auto varTriple = [](const Variable& var) {
+    return TripleComponent{var.name()};
+  };
+  auto varOrTerm = [&varTriple, &graphTerm](VarOrTerm varOrTerm) {
+    return varOrTerm.visit(
+        ad_utility::OverloadCallOperator{varTriple, graphTerm});
+  };
+
+  auto varPath = [](const Variable& var) {
+    return PropertyPath::fromVariable(var);
+  };
+  auto path = [](const PropertyPath& path) { return path; };
+  auto varOrPath = [&varPath,
+                    &path](ad_utility::sparql_types::VarOrPath varOrPath) {
+    return std::visit(ad_utility::OverloadCallOperator{varPath, path},
+                      std::move(varOrPath));
+  };
+
+  auto registerIfVariable = [this](auto& variant) {
+    if (holds_alternative<Variable>(variant)) {
+      visibleVariables_.back().emplace_back(std::get<Variable>(variant));
+    }
+  };
+
+  auto convertTriple = [&varOrTerm, &varOrPath, &registerIfVariable](
+                           TripleWithPropertyPath&& triple) -> SparqlTriple {
+    registerIfVariable(triple.subject_);
+    registerIfVariable(triple.predicate_);
+    registerIfVariable(triple.object_);
+
+    return {varOrTerm(std::move(triple.subject_)),
+            varOrPath(std::move(triple.predicate_)),
+            varOrTerm(std::move(triple.object_))};
+  };
+
+  vector<SparqlTriple> triples = ad_utility::transform(
+      ad_utility::flatten(
+          visitVector<vector<TripleWithPropertyPath>>(ctx->triplesBlock())),
+      convertTriple);
```

Comment:
Without looking at this in detail (we already discussed that you are currently ordering them in the wrong way):

Shouldn't the conversion of the triples (which is currently quite some code, due to issues in the rest of qlever) be in the `triplesBlock` rule, then both functions have a reasonable size.